// ngx-bootstrap/modal/esm2015/models/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function ClassName() { }
function ClassName_tsickle_Closure_declarations() {
    /** @type {?} */
    ClassName.prototype.SCROLLBAR_MEASURER;
    /** @type {?} */
    ClassName.prototype.BACKDROP;
    /** @type {?} */
    ClassName.prototype.OPEN;
    /** @type {?} */
    ClassName.prototype.FADE;
    /** @type {?} */
    ClassName.prototype.IN;
    /** @type {?} */
    ClassName.prototype.SHOW;
}
/**
 * @record
 */
export function Selector() { }
function Selector_tsickle_Closure_declarations() {
    /** @type {?} */
    Selector.prototype.DIALOG;
    /** @type {?} */
    Selector.prototype.DATA_TOGGLE;
    /** @type {?} */
    Selector.prototype.DATA_DISMISS;
    /** @type {?} */
    Selector.prototype.FIXED_CONTENT;
}
/**
 * @record
 */
export function TransitionDurations() { }
function TransitionDurations_tsickle_Closure_declarations() {
    /** @type {?} */
    TransitionDurations.prototype.MODAL;
    /** @type {?} */
    TransitionDurations.prototype.BACKDROP;
}
/**
 * @record
 */
export function DismissReasons() { }
function DismissReasons_tsickle_Closure_declarations() {
    /** @type {?} */
    DismissReasons.prototype.BACKRDOP;
    /** @type {?} */
    DismissReasons.prototype.ESC;
}
