// ngx-bootstrap/timepicker/esm2015/timepicker.models.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function Time() { }
function Time_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    Time.prototype.hour;
    /** @type {?|undefined} */
    Time.prototype.minute;
    /** @type {?|undefined} */
    Time.prototype.seconds;
    /** @type {?|undefined} */
    Time.prototype.isPM;
}
/**
 * @record
 */
export function TimepickerControls() { }
function TimepickerControls_tsickle_Closure_declarations() {
    /** @type {?} */
    TimepickerControls.prototype.canIncrementHours;
    /** @type {?} */
    TimepickerControls.prototype.canIncrementMinutes;
    /** @type {?} */
    TimepickerControls.prototype.canIncrementSeconds;
    /** @type {?} */
    TimepickerControls.prototype.canDecrementHours;
    /** @type {?} */
    TimepickerControls.prototype.canDecrementMinutes;
    /** @type {?} */
    TimepickerControls.prototype.canDecrementSeconds;
    /** @type {?} */
    TimepickerControls.prototype.canToggleMeridian;
}
/**
 * @record
 */
export function TimepickerComponentState() { }
function TimepickerComponentState_tsickle_Closure_declarations() {
    /** @type {?} */
    TimepickerComponentState.prototype.min;
    /** @type {?} */
    TimepickerComponentState.prototype.max;
    /** @type {?} */
    TimepickerComponentState.prototype.hourStep;
    /** @type {?} */
    TimepickerComponentState.prototype.minuteStep;
    /** @type {?} */
    TimepickerComponentState.prototype.secondsStep;
    /** @type {?} */
    TimepickerComponentState.prototype.readonlyInput;
    /** @type {?} */
    TimepickerComponentState.prototype.disabled;
    /** @type {?} */
    TimepickerComponentState.prototype.mousewheel;
    /** @type {?} */
    TimepickerComponentState.prototype.arrowkeys;
    /** @type {?} */
    TimepickerComponentState.prototype.showSpinners;
    /** @type {?} */
    TimepickerComponentState.prototype.showMeridian;
    /** @type {?} */
    TimepickerComponentState.prototype.showSeconds;
    /** @type {?} */
    TimepickerComponentState.prototype.meridians;
}
/**
 * @record
 */
export function TimeChangeEvent() { }
function TimeChangeEvent_tsickle_Closure_declarations() {
    /** @type {?} */
    TimeChangeEvent.prototype.step;
    /** @type {?} */
    TimeChangeEvent.prototype.source;
}
