// ngx-bootstrap/typeahead/esm5/typeahead-match.class.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var TypeaheadMatch = /** @class */ (function () {
    // tslint:disable-next-line:no-any
    function TypeaheadMatch(item, value, header) {
        if (value === void 0) { value = item; }
        if (header === void 0) { header = false; }
        this.item = item;
        this.value = value;
        this.header = header;
    }
    /**
     * @return {?}
     */
    TypeaheadMatch.prototype.isHeader = /**
     * @return {?}
     */
    function () {
        return this.header;
    };
    /**
     * @return {?}
     */
    TypeaheadMatch.prototype.toString = /**
     * @return {?}
     */
    function () {
        return this.value;
    };
    return TypeaheadMatch;
}());
export { TypeaheadMatch };
function TypeaheadMatch_tsickle_Closure_declarations() {
    /** @type {?} */
    TypeaheadMatch.prototype.value;
    /** @type {?} */
    TypeaheadMatch.prototype.item;
    /** @type {?} */
    TypeaheadMatch.prototype.header;
}
