// ngx-bootstrap/pagination/esm2015/models/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function ConfigModel() { }
function ConfigModel_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    ConfigModel.prototype.align;
    /** @type {?} */
    ConfigModel.prototype.boundaryLinks;
    /** @type {?} */
    ConfigModel.prototype.directionLinks;
    /** @type {?} */
    ConfigModel.prototype.firstText;
    /** @type {?} */
    ConfigModel.prototype.itemsPerPage;
    /** @type {?} */
    ConfigModel.prototype.lastText;
    /** @type {?} */
    ConfigModel.prototype.maxSize;
    /** @type {?} */
    ConfigModel.prototype.nextText;
    /** @type {?} */
    ConfigModel.prototype.pageBtnClass;
    /** @type {?} */
    ConfigModel.prototype.previousText;
    /** @type {?} */
    ConfigModel.prototype.rotate;
    /* TODO: handle strange member:
    [key: string]: string | number | boolean;
    */
}
/**
 * @record
 */
export function PagesModel() { }
function PagesModel_tsickle_Closure_declarations() {
    /** @type {?} */
    PagesModel.prototype.text;
    /** @type {?} */
    PagesModel.prototype.number;
    /** @type {?} */
    PagesModel.prototype.active;
}
/**
 * @record
 */
export function PagerModel() { }
function PagerModel_tsickle_Closure_declarations() {
    /** @type {?} */
    PagerModel.prototype.itemsPerPage;
    /** @type {?} */
    PagerModel.prototype.previousText;
    /** @type {?} */
    PagerModel.prototype.nextText;
    /** @type {?} */
    PagerModel.prototype.pageBtnClass;
    /** @type {?} */
    PagerModel.prototype.align;
}
